### ISSUE TITLE
Error type redesign

### ISSUE THREAD

**Issue by seanmonstar** [hyperium/hyper#1131]
The `Error` type in hyper is currently an `enum`, wrapping up the various cases of things that could go wrong when parsing incoming HTTP data. Some of the problems with the current design:

- Being an `enum` means that adding variants is a breaking change. There is a `__Nonexhaustive` variant to try to warn that exhaustive matches can cause breaking changes, but that it's better when internals are hidden.
- Users sometimes try to create `hyper::Error`s, even though they really shouldn't. It is meant only for saying "this thing went wrong while hyper was parsing HTTP", and nothing else. Part of this can be fixed also by changing the actual error types used, such as in outgoing `Stream`s.

Instead, I'd like to propose making the `Error` an opaque `struct`, with some methods to inspect the type and details of an error, somewhat along the lines of `std::io::Error`.

(Still thinking through the specific code, will add a design section later).

**Comment by mp4096** [comment]
Have you considered using [`error-chain`](https://github.com/brson/error-chain)? TBH I didn't use it before, but [it sounds](https://brson.github.io/error-chain/error_chain/index.html#principles-of-error-chain) a lot like what you want to do.

Anyway, maybe @brson himself could say if `error-chain` might be helpful in your case?

**Comment by KodrAus** [comment]
Can you declare the `ErrorKind` enum in `error-chain` to be private? `error-chain` does have the nice benefit of collecting a backtrace from the error's source.

**Comment by seanmonstar** [comment]
A backtrace is neat (but expensive?). I'm not sure I'd want to expose any `enum` at all with regards to the `Error` or `ErrorKind`, because of the inability to grow it. I'd reconsider when and if Rust ever grows some form of private variants (or however is used to prevent exhaustive matches).

It's worth thinking about the things some would do with an `Error` from hyper.

- Just log it. So it needs `fmt::Debug` and `fmt::Display`.
- Try to send a proper `Response` depending on the error that occurred? Like, return `431` when there was a `Error::TooManyHeaders`, perhaps a `414` if there were `Error::TooLarge`, and a `400` with other parse errors.
- Maybe check for certain `io::Error`s, to allow an application to adapt, such as if the error involves too many file descriptors open, or something.

Anything else?

I'm thinking of having methods on the type to allow inspecting the type of error it was, since it's always possible to add more methods. Like, `error.is_uri_too_long()` or whatever.

**Comment by sanmai-NL** [comment]
@seanmonstar: I do not think it is a problem if programmers do exhaustive matching on the `enum` that implements `Error`. Yes, their code will need some small fixes, but only if they upgrade their `hyper` dep. If a new `Error` variant can be returned and exhaustive matching is done, you probably have to alter the logic in your own code anyway, no?

I use `Error` enums extensively i(n closed source code),  I did not encounter ergonomic or API stability issues.

**Comment by KodrAus** [comment]
@sanmai-NL That's interesting, are you able to share a bit about what you do with specific `hyper::Error` variants? In the code I've written using `hyper` I don't think I've had a case where I could do something useful with any specific  variant of `Error`, just shrug and pass it along. But that could definitely just be me so I'm interested to know how other people hold it.

**Comment by sanmai-NL** [comment]
No sorry, I do nothing with `hyper::Error` variants, did you understand my comment as if I do? I do retain the `Error` value as the `cause` field (of struct variants of my `Error` enum) for my own backtraces. What I am trying to point out, is that having one `Error` type to cover lots of code (e.g. current `hyper::Error` as well as the proposed split into server and client errors) seems less productive to me than having specific `Error` types. The latter can easily remain enums, since they are specific to some function scope and only break exhaustive matches on values of that specific `Error` type (return values of that particular function).

**Comment by dtolnay** [comment]
> That's interesting, are you able to share a bit about what you do with specific `hyper::Error` variants?

Maybe translate error messages into another language.

**Comment by sanmai-NL** [comment]
Another general comment: in error handling the viewpoint of regular people needs to be taken into account. I have found this requires tight developer control over error reporting. Backtraces etc. are not very user friendly (may even instill fear), and leak information about the source code, which from a business perspective is intellectual property. For `hyper`, the user faces the web content and I think hyper has an adequate API to return HTTP status codes and nice error pages.

**Comment by KodrAus** [comment]
> Maybe translate error messages into another language.

@dtolnay Oh that's interesting... I've never thought of translating errors before. I guess you could do that on an arbitrary text stream though, rather than individual error variants?

> What I am trying to point out is that having one Error type to cover lots of code ... seems less productive to me than having specific Error types.

@sanmai-NL I have the opposite opinion on this with respect to libraries like `hyper` that are consumed by other libraries. I think having multiple error types that span a single use-case adds boilerplate when trying to chain errors, and cognitive load when consuming them. Now instead of one error type to deal with you've got potentially many. Since the server and client use-cases are distinct splitting them seems reasonable to me.

For application code though I agree having an individual error type per operation is nicer to manage.

> Backtraces etc. are not very user friendly (may even instill fear), and leak information about the source code ...

I think that's a good point. Returning backtraces in responses isn't very good. But having backtraces in logging is very good, and makes issues much easier to diagnose. This is especially true in Rust where you could be handling an error downstream from where it was produced, so the backtrace you get out-of-the-box might be really unhelpful. Maybe for `hyper` having a backtrace somewhere in its internals isn't any more useful than a backtrace where you receive it? But that could be made a concern for application logging rather than on error types directly.

As far as mapping a `hyper` response to a http response in a server goes I thought that would be done by `hyper` directly, rather than by callers? If that's the case then the ergonomics of inspecting the error don't matter so much. If not then maybe we should look at how that code would change with this new design and see if it becomes clunky.

**Comment by seanmonstar** [comment]
> I do not think it is a problem if programmers do exhaustive matching on the enum that implements Error.

It is a bit of a problem. With exhaustive matching, that means that adding a new error case is a **breaking change**. For instance, if at some point it was desirable to separate the `Error::TooLarge` into a `UriTooLong` and `HeaderValueTooLong`, that would break other people's code from compiling. If hyper were at 1.0, that would be mean needing to go to 2.0 just because we changed an internal detail of tracking errors.

> Maybe translate error messages into another language.

A possibility is `Error::code() -> u64` (maybe a smaller number type, probably not going to have trillions of error codes). As a number, you'd have to treat the case of future codes specifically, so adding new ones should be fine.

> Backtraces etc. are not very user friendly (may even instill fear), and leak information about the source code

This definitely true if someone were to print the a `hyper::Error` directly back on a webpage. A worthy thing to consider regarding the use of backtraces in hyper.

> What I am trying to point out is that having one Error type to cover lots of code seems less productive to me than having specific Error types.

That's actually part of the point of this issue. There are 2 rather distinct cases of failure in when using a `Service` (Server or Client), and using `hyper::Error` in both cases is odd.

**Comment by seanmonstar** [comment]
Relevant to the design of type, from [this comment](https://github.com/hyperium/hyper/issues/1130#issuecomment-314529809):

> One thing I'd like is to differentiate IO errors based on where they came from. I care about connect errors in particular [...]

In terms of internal variants, things I see that would be useful to group are:

- Connect
- Parse
- Io

**Comment by nicoburns** [comment]
> It is a bit of a problem. With exhaustive matching, that means that adding a new error case is a breaking change. For instance, if at some point it was desirable to separate the Error::TooLarge into a UriTooLong and HeaderValueTooLong, that would break other people's code from compiling. If hyper were at 1.0, that would be mean needing to go to 2.0 just because we changed an internal detail of tracking errors.

Wouldn't this be a breaking change regardless of whether hyper's error type is an enum or not? If it was a struct that could be inspected somehow, it would still potentially break peoples code, just at runtime rather than at compile time, no?

**Comment by seanmonstar** [comment]
> If it was a struct that could be inspected somehow, it would still potentially break peoples code, just at runtime rather than at compile time, no?

@nicoburns not necessarily. If it were a struct you could inspect, that'd probably be some method, like `err.is_too_large()`. That method can still return the correct value if internally the variant is split into more specific kinds, and new `err.is_uri_too_large()` and `err.is_header_too_large()` methods were added.

The current enum design makes it impossible to add any detail to a variant, since that requires changing the types of values that can matched publicly, so this extra detail could never be provided without a breaking change.

**Comment by algermissen** [comment]
+1 for the code based solution - then there can be a static lookup mechanism (array) to go from code to message. And that would be compatible.

**Comment by seanmonstar** [comment]
Something that I've wanted that is easily done in class-based languages is to be able to split up an error variant into sub-variants in the future, and not breaking existing code.

### With Classes (not Rust)

For example, say there was a `Connect` variant, and later on, we want to split that into smaller pieces, `Resolve` and `Handshake`. With classes, this is easily done. Consider the following Python:

```python
try:
    print(client.get("https://hyper.rs"))
except error.Connect:
    # maybe retry
```

Splitting the errors:

```python
class Resolve(error.Connect):
    pass
class Handshake(error.Connect):
    pass
```

And the original exception code will continue to work, and anyone upgrading *can* update to handle the two cases individually, or continue to just treat them all as `Connect` errors.

## In Rust

In Rust, we don't have classes that can extend like this, so we must use a different strategy to try to achieve this forward compatibility.

### With Methods

If all this information is in private fields, and a user can only ask questions with methods, it is possible.

```rust
if err.is_connect() {
    // retry?
}
```

Adding `is_resolve()` and `is_handshake()` allows futures versions to identify sub-errors, and the `is_connect()` method still works too.

The downside here is that it doesn't work very nicely with pattern matching.

### With Enums

We *could* do this and support pattern matching, but it could get verbose. With a `Kind` enum that you could ask the error about, every variant would need to include a sub-kind (and any sub-kind would additionally need to hold a potentially empty but opaque value to allow for future splitting too. For example:

```rust
if let Err(err) = await client.get(url) {
    match err.kind() {
        ErrorKind::Connect(_) => {
            // retry?
        },
        _ => {},
    }
}
```

If we split the `Connect` error into sub-errors:

```rust
#[non_exhaustive]
enum ErrorKind {
    Connect(Connect),
}

#[non_exhaustive]
enum Connect {
    Resolve(Resolve),
    Handshake(Handshake),
}

struct Resolve(());
struct Handshake(());
```

Then, the original code still works, because the top type is still `ErrorKind::Connect`, and new users can check the `Connect` if they like. It does mean the names can get long, though...

```rust
if let Err(err) = await client.get(url) {
    match err.kind() {
        ErrorKind::Connect(Connect::Resolve(_)) => {
            // don't retry?
        },
        ErrorKind::Connect(Connect::Handshake(_)) => {
            // retry?
        },
        _ => {},
    }
}
```

**Comment by dekellum** [comment]
Regarding just a naming aspect in this design. Chrono has a [similar use of the same names for both a variant and nested sub-type](https://docs.rs/chrono/0.4.0/chrono/format/enum.Item.html).  There at least, it can be challenging to get the  `use` imports right without naming ambiguity. If there are natural names that avoid same-names, that can be friendlier, e.g:

~~~ rust
enum Connect {
    OnResolve(Resolve),
    OnHandshake(Handshake),
}
~~~

**Comment by seanmonstar** [comment]
@dekellum You mean if trying to import the variants as well? `use Connect::*;`?

**Comment by dekellum** [comment]
yes, to make it more compact.

**Comment by sanmai-NL** [comment]
I hope all avoid glob imports, but every importable item should have a unique name irrespective of that.

**Comment by tobz1000** [comment]
Has the (unstable) [#[non_exhaustive]](https://github.com/rust-lang/rfcs/pull/2008) attribute been considered to address the problem of breaking changes?

IMO this is preferable solution, since as a user I would like the be confident that I'm handling every possible error type. This attribute allows for a lint which would let you know when a new error type has been added, and should be handled, without causing a compiler error.

It doesn't address the issue of users creating their own errors, but that seems like a lesser concern to me.

**Comment by sfackler** [comment]
Unless something's changed since the RFC, `#[non_exhaustive]` generates a hard error, not a warning if you try to exhaustively match against of an enum outside of the crate it's defined in.

**Comment by tobz1000** [comment]
I was referring to the possibility of a Clippy lint, or compiler warning in future, which would appear after a minor update in which new variants had been added (see the "Unresolved questions" section).

**Comment by dekellum** [comment]
That would definitely make the feature more pragmatic, less dogmatic, and better for users.

**Comment by dekellum** [comment]
I've started doing [this](https://docs.rs/body-image/0.3.0/body_image/barc/enum.BarcError.html#variant._FutureProof) e.g. a public, documented, unused `_FutureProof` variant because I think the alternative approaches are dysfunctional at best.  I guess only time will tell if I can avoid criticism for breaking changes (e.g. adding error variants) with this approach.

**Comment by rolftimmermans** [comment]
I came here via #1652. I think that methods `is_*` on an `Error` struct are very unergonomic:

* Code that needs to handle many scenarios will be a long `if/else if` chain.
* It is very easy to forget to add net `is_*` methods when the internal enum is extended, leading to holes in the API.
* It is difficult to estimate what level of granularity a user of the library might need to recover from certain errors. The `is_*` methods might be too granular or not granular enough for their use cases.

If the number of internal errors is large and there is a concern for frequent breaking changes, I am greatly in favour of a two-stage classification. Use one enum for the general error category, and other enums for more granular, specific errors. At the very least the general category should be made public so an API consumer can match on it.

**Comment by BeatButton** [comment]
I've recently been having a hard time trying to find a way to get to the actual cause of an error that Hyper gives me. For example, when I receive an error that looks like this:
```
Error(
    Hyper(
        Error(
            Io,
            Os {
                code: 10054,
                kind: ConnectionReset,
                message: "An existing connection was forcibly closed by the remote host.",
            },
        ),
    ),
    "{url goes here}",
)
```
I would like to be able to, somehow, extract the `io::Error` to match against `io::ErrorKind::ConnectionReset`, since that's a specific case I'd like to deal with. I don't think this is possible with the current `Error` type, or at least, I haven't been able to figure out how to do it.

**Comment by sfackler** [comment]
@BeatButton you can use https://doc.rust-lang.org/nightly/std/error/trait.Error.html#method.source or https://docs.rs/hyper/0.12.30/hyper/error/struct.Error.html#method.into_cause

**Comment by keepsimple1** [comment]
Is it still the case that users of this crate should not create `hyper::Error`?

**Comment by Blub** [comment]
It would be nice if instead of the current `into_cause(self) -> Option<...>` we could get something with a signature of `(self) -> Result<Box<dyn ...>, hyper::Error>` for when we really only want to catch eg. an `io::Error` and pass the rest through, similar to downcast methods, where the other kinds are simply returned unchanged as the `Err`. Currently you're forced to check all the methods (with no warnings or errors if new methods are added in the future) before then finally doing an `into_cause`.
(Or just get the enum exposed... because methods make for really ugly error handling code)

**Comment by aagmv** [comment]
I want to distinguish errors that happen during a long-running transfer of the body, after headers have been sent, from other errors, e.g. via the `BodyWrite` ErrorKind.
But the public API doesn't expose this information and I had to approximate it by checking for `BrokenPipe` IO errors in the source-chain instead, which is brittle and less portable.

Previously I also considered doing a 100-Continue dance instead to delineate phases (and thus where the error occurred), but it turns out that `reqwest` doesn't support waiting for that preliminary response either. #3833 and https://github.com/seanmonstar/reqwest/issues/2845

So generally which phase the sending and receiving halves were in when an error occurred would be useful as high-level information.

**Comment by seanmonstar** [comment]
The `Error` type was redesigned into an opaque struct before we went to 1.0. So, the original motivation is complete. There's some discussion in here, and also some other issues about improving that type, but it's for the best to close this one out.
